### PR TITLE
Adds tracking for failed tests, re-running failed tests.

### PR DESCRIPTION
**Pull request by rbuckton**
This change adds the following capabilities:
* After running tests (either via `runtests` or `runtests-parallel`), any failing tests will be written to a _.failed-tests_ file.
  * This file is in a response-file format understood by mocha's `--opts` option.
  * The names of failing tests are escaped and combined to be used with mocha's `--grep` option.
* Running either `jake runtests failed=true` or `gulp runtests --failed` will run only the failing tests from the last test run.
  * The `failed` option will be superseded by the `tests` option, if both are provided.
  * The `failed` option is not currently supported on the `runtests-parallel` task.
* When a test run completes successfully, the _.failed-tests_ file is removed unless you also pass `keepFailed=true` to **jake** or `--keepFailed` to **gulp**.

**Examples (jake)**
```
# Run tests with test failure tracking
jake runtests
jake runtests-parallel

# Run failed tests (remove .failed-tests file on success)
jake runtests failed=true

# Run failed tests (do not remove .failed-tests file on success)
jake runtests failed=true keepFailed=true
```

**Examples (gulp)**
```
# Run tests with test failure tracking
gulp runtests
gulp runtests-parallel

# Run failed tests (remove .failed-tests file on success)
gulp runtests --failed

# Run failed tests (do not remove .failed-tests file on success)
gulp runtests --failed --keepFailed
```

### CODE REVIEW

**Comment by yuit** [comment]
Gotta take a look but this is great :)

File: scripts/mocha-file-reporter.js [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    }
+}
+
+/**
+ * @param file {String}
+ * @param failures {Array}
+ * @param keepFailed {Boolean}
+ * @param fn {Function}
+ */
+FileReporter.writeFailures = function (file, failures, keepFailed, fn) {
```
Comment (yuit) [inline scripts/mocha-file-reporter.js]: Could you rename `file` to `fileName` or `filePath`

File: Gulpfile.ts [hunk 2]
```diff
@@ -57,9 +58,12 @@ const cmdLineOptions = minimist(process.argv.slice(2), {
         inspect: process.env.inspect || process.env["inspect-brk"] || process.env.i,
         host: process.env.TYPESCRIPT_HOST || process.env.host || "node",
         browser: process.env.browser || process.env.b || "IE",
+        bail: process.env.bail || false,
```
Comment (rbuckton) [inline Gulpfile.ts]: Exit on the first failed test. This already exists in _jakefile.js_

File: scripts/mocha-file-reporter.js (hunk 1 above)
Comment (rbuckton) [inline scripts/mocha-file-reporter.js]: I only added it here for better intellisense.

File: scripts/mocha-file-reporter.js [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+exports = module.exports = FileReporter;
+
+function FileReporter(runner, options) {
+    if (!runner) return;
+    options = options || {};
+
+    var reporterOptions = this.reporterOptions = options.reporterOptions || {};
+    reporterOptions.file = reporterOptions.file || ".failed-tests";
+    reporterOptions.keepFailed = reporterOptions.keepFailed || false;
+    reporterOptions.reporter = reporterOptions.reporter;
```
Comment (sandersn) [inline scripts/mocha-file-reporter.js]: this looks like a rigid idiom, but it still bugs me. Can you remove it?

**Comment by mhegazy** [comment]
@rbuckton can you refresh this PR?

**Comment by rbuckton** [comment]
Superseded by #25004.

**Pull request closed without merging**